permission/eea/backend.go

```diff
@@ -0,0 +1,348 @@
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	eb "github.com/ethereum/go-ethereum/permission/eea/bind"
+	ptype "github.com/ethereum/go-ethereum/permission/types"
+)
+
+type Backend struct {
+	Ib    ptype.InterfaceBackend
+	Contr *Init
+}
+
+func (b *Backend) ManageAccountPermissions() error {
+	chAccessModified := make(chan *eb.AcctManagerAccountAccessModified)
+	chAccessRevoked := make(chan *eb.AcctManagerAccountAccessRevoked)
+	chStatusChanged := make(chan *eb.AcctManagerAccountStatusChanged)
+
+	opts := &bind.WatchOpts{}
+	var blockNumber uint64 = 1
+	opts.Start = &blockNumber
+
+	if _, err := b.Contr.PermAcct.AcctManagerFilterer.WatchAccountAccessModified(opts, chAccessModified); err != nil {
+		return fmt.Errorf("failed AccountAccessModified: %v", err)
+	}
+
+	if _, err := b.Contr.PermAcct.AcctManagerFilterer.WatchAccountAccessRevoked(opts, chAccessRevoked); err != nil {
+		return fmt.Errorf("failed AccountAccessRevoked: %v", err)
+	}
+
+	if _, err := b.Contr.PermAcct.AcctManagerFilterer.WatchAccountStatusChanged(opts, chStatusChanged); err != nil {
+		return fmt.Errorf("failed AccountStatusChanged: %v", err)
+	}
+
+	go func() {
+		stopChan, stopSubscription := ptype.SubscribeStopEvent()
+		defer stopSubscription.Unsubscribe()
+		for {
+			select {
+			case evtAccessModified := <-chAccessModified:
+				types.AcctInfoMap.UpsertAccount(evtAccessModified.OrgId, evtAccessModified.RoleId, evtAccessModified.Account, evtAccessModified.OrgAdmin, types.AcctStatus(int(evtAccessModified.Status.Uint64())))
+
+			case evtAccessRevoked := <-chAccessRevoked:
+				types.AcctInfoMap.UpsertAccount(evtAccessRevoked.OrgId, evtAccessRevoked.RoleId, evtAccessRevoked.Account, evtAccessRevoked.OrgAdmin, types.AcctActive)
+
+			case evtStatusChanged := <-chStatusChanged:
+				if ac, err := types.AcctInfoMap.GetAccount(evtStatusChanged.Account); ac != nil {
+					types.AcctInfoMap.UpsertAccount(evtStatusChanged.OrgId, ac.RoleId, evtStatusChanged.Account, ac.IsOrgAdmin, types.AcctStatus(int(evtStatusChanged.Status.Uint64())))
+				} else {
+					log.Info("error fetching account information", "err", err)
+				}
+			case <-stopChan:
+				log.Info("quit account contract watch")
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+func (b *Backend) ManageRolePermissions() error {
+	chRoleCreated := make(chan *eb.RoleManagerRoleCreated, 1)
+	chRoleRevoked := make(chan *eb.RoleManagerRoleRevoked, 1)
+
+	opts := &bind.WatchOpts{}
+	var blockNumber uint64 = 1
+	opts.Start = &blockNumber
+
+	if _, err := b.Contr.PermRole.RoleManagerFilterer.WatchRoleCreated(opts, chRoleCreated); err != nil {
+		return fmt.Errorf("failed WatchRoleCreated: %v", err)
+	}
+
+	if _, err := b.Contr.PermRole.RoleManagerFilterer.WatchRoleRevoked(opts, chRoleRevoked); err != nil {
+		return fmt.Errorf("failed WatchRoleRemoved: %v", err)
```

Comment:
Minor typo in error message - should be "WatchRoleRevoked

Reply:
done